### PR TITLE
Specify allowed characters for Baggage keys and values

### PR DESCRIPTION
**Pull request by yurishkuro**
## Changes

* Add missing spec language on what keys and values are allowed for the baggage.
* Emphasize case sensitivity
* Definitions are aligned with W3C Baggage Spec, but provide a distinction between API representation and the wire encoding (done by W3C Propagator)

### CODE REVIEW

**Review by reyang — DISMISSED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,7 +7,9 @@ release.
 
 ## Unreleased
 
-### Context
+### Context & Baggage
+
+- Specify allowed characters for Baggage keys and values.
```
Comment (reyang) [inline CHANGELOG.md:12]:
```suggestion
- Specify allowed characters for Baggage keys and values.
([#3801](https://github.com/open-telemetry/opentelemetry-specification/pull/3801))
```

**Comment by yurishkuro** [comment]
Note: this could be a breaking change e.g. for Go SDK (https://github.com/open-telemetry/opentelemetry-go/issues/3685), whose current implementation is wrong imo.

File: specification/baggage/api.md [hunk 2]
```diff
@@ -29,6 +29,19 @@ metrics, traces, and logs. It is a set of name/value pairs describing
 user-defined properties. Each name in `Baggage` MUST be associated with
 exactly one value.
 
+Baggage **names** are strings that satisfy the `token` definition from
+[RFC7230, Section 3.2.6](https://tools.ietf.org/html/rfc7230#section-3.2.6).
+Alpha-numeric names are strongly recommended. Language API MAY return
```
Comment (pellared) [inline specification/baggage/api.md:34]: nit

```suggestion
Alpha-numeric names are strongly recommended. Language API SHOULD return
```

**Review by pellared — APPROVED** [review]
> Note: this could be a breaking change e.g. for Go SDK ([open-telemetry/opentelemetry-go#3685](https://github.com/open-telemetry/opentelemetry-go/issues/3685)), whose current implementation is wrong imo.

I would say that we chose the wrong decision in OTel Go. Still we might try to fix it in some backwards-compatible way (by extending the API). 

TL;DR; For me this is not a blocker for this PR.

**Comment by pellared** [comment]
I am working towards making a fix in OTel Go: https://github.com/open-telemetry/opentelemetry-go/pull/4804

File: specification/baggage/api.md [hunk 3]
```diff
@@ -29,6 +29,19 @@ metrics, traces, and logs. It is a set of name/value pairs describing
 user-defined properties. Each name in `Baggage` MUST be associated with
 exactly one value.
 
+Baggage **names** are strings that satisfy the `token` definition from
```
Comment (dyladan) [inline specification/baggage/api.md:32]: The W3C is currently considering allowing UTF-8 strings for keys as well as values. The otel API can make their own decision, but I think the UTF-8 change is likely to go through in W3C at this point.

Comment (yurishkuro) [reply specification/baggage/api.md:32]: Please link a ticket. Such change in W3C will be seriously breaking.

**Comment by jkwatson** [comment]
Java is also not compliant with this. Currently, we only propagate keys/values that meeting this definition, but we allow arbitrary keys for in-memory usage.

File: specification/baggage/api.md [hunk 4]
```diff
@@ -35,6 +35,19 @@ with _exactly one value_. This is more restrictive than the [W3C Baggage
...
 which allows duplicate entries for a given name.
 
+Baggage **names** are strings that satisfy the `token` definition from
+[RFC7230, Section 3.2.6](https://tools.ietf.org/html/rfc7230#section-3.2.6).
+Alpha-numeric names are strongly recommended. Language API SHOULD return
+an error when given a baggage name that is not a `token`.
+
+Baggage **values** are any valid UTF-8 strings. Language API MUST accept
+any valid UTF-8 string as baggage value. Language API MUST guarantee that
+`Get` and `Set` operations are symmetrical with respect to name and value,
```
Comment (ocelotl) [inline specification/baggage/api.md:45]:
```suggestion
```

File: specification/baggage/api.md [hunk 5]
```diff
@@ -35,6 +35,19 @@ with _exactly one value_. This is more restrictive than the [W3C Baggage
 Specification, § 3.2.1.1](https://www.w3.org/TR/baggage/#baggage-string)
 which allows duplicate entries for a given name.
 
+Baggage **names** are strings that satisfy the `token` definition from
+[RFC7230, Section 3.2.6](https://tools.ietf.org/html/rfc7230#section-3.2.6).
+Alpha-numeric names are strongly recommended. Language API SHOULD return
```
[the remark this answers is not in the document]
Comment (yurishkuro) [reply specification/baggage/api.md:40]: SHOULD is a normative word that does not apply to user-supplied values, it's not something that the OTEL SDK can control.

File: specification/baggage/api.md (hunk 3 above)
Comment (yurishkuro) [reply specification/baggage/api.md:38]: https://github.com/w3c/baggage/issues/99

There is no resolution there as of Jan 2024. I don't think it should block this PR - it makes changes to match the __current__ W3C spec. If we adopt ahead of time more extensive alphabet for keys, then OTEL simply won't be compatible w/ W3C spec.

File: specification/baggage/api.md [hunk 6]
```diff
@@ -35,6 +35,19 @@ with _exactly one value_. This is more restrictive than the [W3C Baggage
...
+an error when given a baggage name that is not a `token`.
+
+Baggage **values** are any valid UTF-8 strings. Language API MUST accept
+any valid UTF-8 string as baggage value. Language API MUST guarantee that
```
[the remark this answers is not in the document]
Comment (yurishkuro) [reply specification/baggage/api.md]: phrased it differently

File: specification/baggage/api.md (hunk 4 above)
Comment (yurishkuro) [reply specification/baggage/api.md]: phrased it differently

File: specification/baggage/api.md [hunk 7]
```diff
@@ -35,6 +35,19 @@ with _exactly one value_. This is more restrictive than the [W3C Baggage
...
+`Get` and `Set` operations are symmetrical with respect to name and value,
+i.e. `Get` MUST return the same **value** string as received in `Set`.
```
[the remark this answers is not in the document]
Comment (yurishkuro) [reply specification/baggage/api.md]: "i.e." is not used to give an example, but to provide alternative wording. But you're right, "symmetrical" was unnecessary, I phrased it in simpler terms, and provided explicit examples (thanks for a good idea!)

**Review by ocelotl — APPROVED** [review]

File: specification/baggage/api.md [hunk 8]
```diff
@@ -35,6 +35,27 @@ with _exactly one value_. This is more restrictive than the [W3C Baggage
 Specification, § 3.2.1.1](https://www.w3.org/TR/baggage/#baggage-string)
 which allows duplicate entries for a given name.
 
+Baggage **names** are strings that satisfy the `token` definition from
+[RFC7230, Section 3.2.6](https://tools.ietf.org/html/rfc7230#section-3.2.6).
+Alpha-numeric names are strongly recommended. Language API SHOULD return
+an error when given a baggage name that is not a `token`.
```
Comment (jack-berg) [inline specification/baggage/api.md:41]: > Java is also not compliant with this. Currently, we only propagate keys/values that meet this definition, but we allow arbitrary keys for in-memory usage.

➕ 

Our stance in java is that the API should never return an error. If anyone uses the API incorrectly, we log a warning and use a no-op implementation of the operation. Applying this logic to baggage and these new key constraints, when a invalid key / value is recorded in baggage, we would log an error and skip recording that key / value pair. Like @jkwatson said, this is not our behavior today and skipping recording those key / value pairs is a breaking behavioral change. But because we only propagate the valid key / values, the consequence of us not enforcing these constraints is only that other consumers of baggage in process (i.e. span processors, log processors) can access key / values that will not be propagated.

@jkwatson I propose that if this is merged, we keep the existing behavior which records the invalid key / values (since changing would be breaking) but add a check which logs if the key / value is invalid. If invalid, log a message indicating that the key / value won't be propagated, but record it for access by in-process consumers.

Comment (yurishkuro) [reply specification/baggage/api.md:41]: I am fine changing SHOULD to MAY

Comment (tedsuo) [reply specification/baggage/api.md:41]: Actually, API methods in OpenTelemetry MUST NOT return errors. A first principle of OpenTelemetry API design is that users should never have to wrap any OpenTelemetry API call in error or exception handling code.

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/error-handling.md#basic-error-handling-principles

File: specification/baggage/api.md (hunk 3 above)
Comment (pellared) [reply specification/baggage/api.md:38]: @dyladan Would the change be breaking? 

As right now a key like: `%20` is valid and will not be encoded. I guess after https://github.com/w3c/baggage/issues/99 it would be still valid but the key will be encoded to a whitespace: ` `.

File: specification/baggage/api.md (hunk 8 above)
Comment (yurishkuro) [reply specification/baggage/api.md:41]: > API methods in OpenTelemetry MUST NOT return errors

they must not raise exceptions, but they _can_ return errors when idiomatic in the language (e.g. Try in Scala)

Comment (yurishkuro) [reply specification/baggage/api.md:41]: Proposed change:

> Language API MUST ignore baggage entries given a name that is not a `token`, e.g. they MAY return an error (if idiomatic for the language) or fallback to a no-op implementation.

Comment (jack-berg) [reply specification/baggage/api.md:41]: > Applying this logic to baggage and these new key constraints, when a invalid key / value is recorded in baggage, we would log an error and skip recording that key / value pair. Like @jkwatson said, this is not our behavior today and skipping recording those key / value pairs is a breaking behavioral change.

Sorry if my message wasn't clear. Invalid key / values aren't propagated across application boundaries, but they are recorded in baggage such that they can be used by other baggage consumers like processors. 

I believe your intent here would be to say that not only should invalid key / value pairs not be propagated, but they also shouldn't be recorded in baggage in the first place. The "not recorded in baggage in the first place" part is what I think is a breaking change for java.

Comment (yurishkuro) [reply specification/baggage/api.md:41]: I see. The question of how encoding restrictions impact in-process API is always a tricky one, especially since encodings technically could vary across propagators. But we want to ensure that at least w3c propagator works, hence the `token` restriction. The main function of baggage is not in-process propagation, but across processes - for in-process people should use Context API directly with no restrictions. The fact that Java Baggage implementation provided accidental feature of unrestricted in-process propagation should not be treated as a breaking change if that behavior is restricted, imo, as long as it still maintains its cross-process propagation capability (the primary function).

File: specification/baggage/api.md (hunk 3 above)
Comment (dyladan) [reply specification/baggage/api.md:38]: Im sorry I didn't follow up. We're working with the i11y group at w3c but it seems this change will in fact not go through in the baggage spec. It is not providing enough value to justify breaking our existing implementations.

Comment (pellared) [reply specification/baggage/api.md:38]: I think this conversation can be resolved.

File: specification/baggage/api.md (hunk 8 above)
Comment (pellared) [inline specification/baggage/api.md:41]: Should the Baggage API have the same constrains as W3C Baggage Propagator? Is this change not mixing different concerns?

The Baggage API can be used intra-process and then usage of UTF-8 keys would not be a problem.

Shouldn't the W3C Baggage Propagator report a problem (e.g. emit some logs) when it encounters a key that it cannot encode?

**Review by pellared — CHANGES_REQUESTED** [review]

File: specification/baggage/api.md (hunk 8 above)
Comment (pellared) [reply specification/baggage/api.md:41]: Should the Baggage API have the same constrains as W3C Baggage Propagator? Is this change not mixing different concerns?

The Baggage API can be used in-process and then usage of UTF-8 keys would not be a problem.

> The main function of baggage is not in-process propagation, but across processes - for in-process people should use Context API directly with no restrictions. 

I will say that is it "one" of the functions. Even the [spec](https://opentelemetry.io/docs/concepts/signals/baggage/) says:

> Baggage MAY be propagated across process boundaries or across any arbitrary boundaries

And [here](https://opentelemetry.io/docs/concepts/signals/baggage/#why-does-otel-baggage-exist):

> Baggage provides a uniform way to store and propagate information across a trace and other signals. For example, you may want to attach information from your application to a span and retrieve that information much later and use it later on with another span.

Shouldn't the W3C Baggage Propagator report a problem (e.g. emit some logs) when it encounters a key that it cannot encode?

**Review by pellared — CHANGES_REQUESTED** [review]
https://github.com/open-telemetry/opentelemetry-specification/pull/3801#discussion_r1471084605

**Comment by carlosalberto** [comment]
I support @pellared comment on different concerns (the in-process API and the inter-process propagation). Maybe we can adjust this PR to only mention that invalid keys MUST NOT be propagated? Then we can spin off a separate issue/PR discussing on `Baggage` API part.

(Also this would be easy to do as it would align with what some SIGs _already_ do)

**Comment by yurishkuro** [comment]
Updated to allow arbitrary strings in the API, but caution that Propagators may not allow some keys.

File: specification/baggage/api.md [hunk 9]
```diff
@@ -35,6 +35,32 @@ with _exactly one value_. This is more restrictive than the [W3C Baggage
 Specification, § 3.2.1.1](https://www.w3.org/TR/baggage/#baggage-string)
 which allows duplicate entries for a given name.
 
+Baggage **names** are any valid UTF-8 strings. Language API SHOULD NOT
+restrict which strings are used as baggage **names**. However, the
+specific `Propagator`s that are used to transmit baggage entries across
+component boundaries may impose their own restrictions on baggage names.
+For example, the [W3C Baggage specification](https://github.com/w3c/baggage/blob/main/baggage/HTTP_HEADER_FORMAT.md#key)
```
Comment (pellared) [inline specification/baggage/api.md:42]: When possible I think we should refer to [W3C Working Draft](https://www.w3.org/standards/types#WD)

```suggestion
For example, the [W3C Baggage specification](https://www.w3.org/TR/baggage/#key)
```

File: specification/baggage/api.md [hunk 10]
```diff
@@ -35,6 +35,32 @@ with _exactly one value_. This is more restrictive than the [W3C Baggage
...
+from [RFC7230, Section 3.2.6](https://tools.ietf.org/html/rfc7230#section-3.2.6).
+For maximum compatibility, alpha-numeric names are strongly recommended
+to be used as baggage names.
+
+Baggage **values** are any valid UTF-8 strings. Language API MUST accept
+any valid UTF-8 string as baggage **value** in `Set` and return the same
+value from `Get`.
+
+Language API MUST treat both baggage names and values as case sensitive.
+See also [W3C Baggage -- Case sensitivity of keys](https://github.com/w3c/baggage/blob/main/baggage/HTTP_HEADER_FORMAT_RATIONALE.md#case-sensitivity-of-keys).
```
Comment (pellared) [inline specification/baggage/api.md:53]: I suggest to clearly say that this is from Rationale docs.

```suggestion
See also [W3C Baggage Rationale](https://github.com/w3c/baggage/blob/main/baggage/HTTP_HEADER_FORMAT_RATIONALE.md#case-sensitivity-of-keys).
```

**Review by pellared — APPROVED** [review]

**Review by carlosalberto — APPROVED** [review]

**Review by reyang — APPROVED** [review]

File: specification/baggage/api.md (hunk 8 above)
Comment (yurishkuro) [reply specification/baggage/api.md]: Changed to allow any strings.

**Comment by yurishkuro** [comment]
cc @open-telemetry/go-maintainers - the new wording makes Go implementation non-compliant, but it's not a breaking change for its API, you will just need to relax the checks (i.e. never return errors from `baggage.New`)

**Comment by carlosalberto** [comment]
Let's remind maintainers in the call next Monday about this, but otherwise we are fine to go.

**Comment by yurishkuro** [comment]
@carlosalberto are we good to merge?
